cdap-data-fabric/src/main/java/io/cdap/cdap/data2/dataset2/lib/cube/DefaultCube.java

```diff
@@ -110,8 +129,9 @@ public void add(CubeFact fact) {
 
   @Override
   public void add(Collection<? extends CubeFact> facts) {
-    List<Fact> toWrite = Lists.newArrayList();
+    Map<List<DimensionValue>, List<Fact>> toWrite = new HashMap<>();
```

Comment:
Now that toWrite is a map, is the calculation of `cube.tsFact.created.count` and `cube.tsFact.added.count` correct at the end of this method? instead of `toWrite.size()`, may be use `numFacts`?

Reply:
Yes, thanks! I forgot to include this change (I have a list of commit to split into PRs)

Reply:
You missed removing the `.size()` call. Otherwise LGTM.

Reply:
Oops, done